### PR TITLE
Use Poetry for dependency management

### PR DESCRIPTION
**Pull request by eloquence**
## Status

Ready for review, but should not be merged until https://github.com/freedomofpress/securedrop-builder/pull/467 and the changes in the `poetry-support-with-proxy-changes` in `securedrop-builder` are merged, after which we can drop the `FIXME` change from the CI config.

## Description
This PR migrates the `securedrop-proxy` package to use Poetry for dependency management, consistent with our overall plan to use Poetry for most Python dependency management going forward. It also updates the documentation accordingly.

CI is red due to #121, which is unrelated to this change. 

## Test plan

The test plan in https://github.com/freedomofpress/securedrop-builder/pull/467 verifies the basic workflow already: It asks the tester to install the dependencies via Poetry, and to add a dependency for testing purposes.

Beyond that, I would recommend testing the modified `Makefile` targets in this PR.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -103,8 +109,18 @@ jobs:
...
+      - when:
+          condition:
+            equal: [ bullseye, << parameters.image >> ]
+          steps:
+            - *install_poetry_bullseye
+      - when:
+          condition:
+            equal: [ bookworm, << parameters.image >> ]
+          steps:
+            - *install_poetry_bookworm
```
Comment (eloquence) [inline .circleci/config.yml:122]: I tried to refactor this into an anchor to DRY it, but CircleCI's config parser wasn't happy with placing conditions in anchors. If you see a way to get that to work, happy to refactor.

File: pyproject.toml [hunk 2]
```diff
@@ -3,3 +3,28 @@ line-length = 100
...
+black = "^23.7.0"
+coverage = "^7.3.0"
+flake8 = "^6.1.0"
+isort = "^5.12.0"
+mypy = "^1.5.1"
+types-PyYAML = "^6.0.12.11"
+types-requests = "^2.31.0.2"
+vcrpy = "^5.1.0"
+bandit = "*"
+safety = "*"
```
Comment (eloquence) [inline pyproject.toml:30]: We previously did not specify this as a test dependency at all, and just `pip install`ed it in the Makefile targets.

I think it's best to enumerate all test dependencies as such. If we want to make sure that we always get the latest version, we can `poetry update` the dependency before each run. That's what the Makefile targets do now.

File: Makefile [hunk 3]
```diff
@@ -3,83 +3,62 @@
...
-			&& safety check --full-report -r $$req_file \
-			&& echo -e '\n' \
-			|| exit 1; \
-		done
+	@echo "Installing latest version of safety in virtual environment…"
+	@poetry update safety
+	@echo "Running safety against requirements files in this repository…"
+	for req_file in `find . -type f -name '*requirements.txt'`; do \
+		echo "Checking file $$req_file" \
+		&& poetry run safety check --full-report -r $$req_file \
```
Comment (eloquence) [inline Makefile:20]: If we keep `safety` around at all, I'd like to of course point it at the Poetry lockfile. Unfortunately that's not possible yet (the good old `category` change broke `safety` as well: https://github.com/pyupio/safety/issues/479), so we do lose some `safety` check functionality here (but we can turn on Dependabot to make up for it).

File: Makefile [hunk 4]
```diff
@@ -3,83 +3,62 @@
...
-		for req_file in `find . -type f -name '*requirements.txt'`; do \
-			echo "Checking file $$req_file" \
-			&& safety check --full-report -r $$req_file \
-			&& echo -e '\n' \
-			|| exit 1; \
-		done
+	@echo "Installing latest version of safety in virtual environment…"
+	@poetry update safety
+	@echo "Running safety against requirements files in this repository…"
+	for req_file in `find . -type f -name '*requirements.txt'`; do \
```
Comment (legoktm) [inline Makefile:18]: Now that there's only one build-requirements.txt file can we get rid of the for+find loop?

**Review by legoktm — CHANGES_REQUESTED** [review]
Nice :)

File: .circleci/config.yml (hunk 1 above)
Comment (eloquence) [reply .circleci/config.yml]: Makes sense, done!

File: pyproject.toml (hunk 2 above)
Comment (eloquence) [reply pyproject.toml:30]: Yeah, upon reflection, I moved this out of the Makefile target and into the CI job at least. IMO it should always be clearly the developer's intent to update.

File: .circleci/config.yml [hunk 5]
```diff
@@ -1,11 +1,25 @@
 ---
 common-steps:
+  - &install_poetry_bullseye
+    run:
+      name: Install Poetry on Debian Bullseye (PyPI package)
+      command: |
+        set -e
+        apt update && apt install -y --no-install-recommends python3-pip
```
[the remark this answers is not in the document]
Comment (eloquence) [reply .circleci/config.yml]: Done!

File: README.md [hunk 6]
```diff
@@ -29,27 +29,40 @@ https://github.com/freedomofpress/securedrop-workstation/issues/107.
...
-source .venv/bin/activate
-make test
-```
+#### Managing Dependencies
 
-#### Update Dependencies
+We use Poetry to manage dependencies for this project.
 
-If you're adding or updating a dependency, you need to:
+### Development dependencies
```
[the remark this answers is not in the document]
Comment (eloquence) [reply README.md:39]: Agreed, will work on that once we've got the first project transitioned over.

File: Makefile (hunk 4 above)
Comment (eloquence) [reply Makefile]: Done!

**Comment by eloquence** [comment]
CI will now fail fully until https://github.com/freedomofpress/securedrop-builder/pull/469 is merged.

**Review by legoktm — APPROVED** [review]
I squashed it down, overall LGTM! Waiting for CI, with the note that bookworm CI is expected to partially fail.

**Pull request merged**